### ISSUE TITLE
File attachments allow illegal path characters

### ISSUE THREAD

**Issue by lakam99** [AdhocAdam/smletsExchangeConnector#324]
**Describe the bug**
We've discovered in our production environment, tickets with file attachments that have **_illegal path characters_**. Typically, this includes "?", " | ", ":". The email that comes into the SMLet managed inbox always comes in as 'message.eml' which is good, but the email's subsequent file attachments come in as the full subject line which usually contains an array of special characters. When someone tries to work on a ticket with one of these erroneous files attached from the Cireson Portal, it prohibits any change. Trying to open the erroneous file attachment returns the error, "Illegal characters in path". Trying to open the erroneous attachment through the 2019 Service Manager Console also returns a fatal error and causes the console to crash and shut down. 

**Help us reproduce the bug**
Version of the connector you are using: **v3.3.1.21**
Features/variables you have enabled (i.e. $enableAzureCognitiveServicesForNewWI):
- Use impersonation
- Create Users not Found in CMDB
- Attach Email to Work Item
- Move to Deleted Items After Processing
- Update Work Items missing [Work Item] in the Email Subject by matching Exchange Conversation ID (requires Attach Email to Work Item)
- If Work Item to Update is closed - Create, Relate, and Copy into a New Work Item.
- Process Digitally Signed emails that have certificate errors (expired, invalid, etc)
- Process Multiple Inboxes
- AI: disabled

**Expected behavior**
All emails that would be attached to a ticket with the subject line as the file name should be cleansed of all illegal path characters.

**Location and Environment**
Where are you running this from? Task Scheduler? SMA? Workflow server?:
Task scheduler on Service Manager Management Server.

What SCSM version are you running? (i.e. 2012R2, 2016, 2016 UR3, 1801, etc.): 
2019 UR2

**Additional context**
To reproduce this bug, create an email with a subject line containing illegal characters such as '?', '|', ':', '%', etc... Send the email to anyone but not to the connector. Then, create another email addressed to the connector mailbox, and attach the email you previously sent using the 'Attach Item > Outlook Item > Sent Items > [Your email with illegal path characters]'.
When the ticket comes in:

- Opening it in the Cireson Portal, then trying to make a change to the ticket, will give you an error regarding how something is not set to an instance of an object. If you go to the 'Related Items' tab and try to open the erroneous email, you'll be redirected to an error page saying the file has illegal characters in its path.
- Opening it in the 2019 Service Manager Console is a bit strange. The _first_ time you open the email from the 'Related Items' tab, it will work. The _second_ and every other time after, Service Manager will crash.

The overall gist of it is, emails with email attachments come in with the subject line as the filename. This causes issues with tickets because the email filename may contain illegal path characters. The solution would to be cleanse filenames of illegal path characters _before_ attaching them to a ticket. Thank you!
![error_screencap](https://user-images.githubusercontent.com/1222447/139479105-f2761df0-6cca-4c16-a6e0-17744b982751.PNG)

**Comment by AdhocAdam** [comment]
Thank you for your unbelievably detailed breakdown here @lakam99. It is beyond appreciated! 👍 

While I can't reproduce the Console error you cite, I can reproduce the error you experience in the Portal.
- Same page error per your screenshot
- Same popup around "...not set to an instance of an object"

Whichever the case, I agree with you that what probably needs to happen here is a cleansing of the file name to prevent this from occurring. I'll update this thread further as I dig in.

**Comment by AdhocAdam** [comment]
I gotta hand it to you. The reproduction step you had of:
> Attach Item > Outlook Item > Sent Items > [Your email with illegal path characters]'

Immediately. **_IMMEDIATLEY_** led me to "What happens in Outlook if I open an email with illegal characters in the Subject and attempt to save it to my desktop?"

![outlookSave03](https://user-images.githubusercontent.com/6636040/139518880-384b257b-5e74-4c16-bc80-8807d23bb96c.PNG)

As it turns out, Outlook's Save As File Prompt pulls the characters from the file name. Which after some quick searching led me to the Microsoft Documentation on the [Path.GetInvalidFileNameChars Method](https://docs.microsoft.com/en-us/dotnet/api/system.io.path.getinvalidfilenamechars?view=net-5.0).

I'm guessing there are probably a few different ways to make use of the method. But as far as the PowerShell is concerned, I threw the subject from above into a string and...

```powershell
$attachmentName = "did you see < | > that: ludicrous display `"last`" night?.msg"
$attachmentName.Split([IO.Path]::GetInvalidFileNameChars()) -join ""
```

Using .Split() and -join appears to produce a result identical to what Outlook does:
<pre>did you see    that ludicrous display last night.msg</pre>

Which means I should be able to update these lines in the Attach-FileToWorkItem function accordingly so that the File Attachment is created correctly from the outset.
https://github.com/AdhocAdam/smletsexchangeconnector/blob/4740de699b759a09f2af15081751f0ba3e67ddc0/smletsExchangeConnector.ps1#L2348

https://github.com/AdhocAdam/smletsexchangeconnector/blob/4740de699b759a09f2af15081751f0ba3e67ddc0/smletsExchangeConnector.ps1#L2352

I have some more testing to do, but so far so good. 😄

**Comment by AdhocAdam** [comment]
So I updated the two lines in the connector above as follows:
```powershell
$NewFile.Item($fileAttachmentClass, "DisplayName").Value = $attachment.Name.Split([IO.Path]::GetInvalidFileNameChars()) -join ""
$NewFile.Item($fileAttachmentClass, "DisplayName").Value = $attachment.DisplayName.Split([IO.Path]::GetInvalidFileNameChars()) -join ""
```

Since Split doesn't throw an error if the character(s) to split on is not present, this is a safe and reliable way to remove illegal characters when the File Attachment is created in SCSM. Also, out of my own curiosity I ran the method in PowerShell to see a host of characters this method references.

![splitChars](https://user-images.githubusercontent.com/6636040/139597358-7111a91b-71e7-4e4f-a3ca-b1874ccde3d3.PNG)
-


So of course the next thing I figured I'd try to do is wire up a quick loop in my lab to:
- identify File Attachments suffering from this naming issue
- identify their related Work Item
- perform a targeted update to remove illegal characters from that File Attachment's DisplayName property

```powershell
#Get all File Attachments, sort starting with the most recent
$fileAttachmentClass = Get-SCSMClass -Name "System.FileAttachment$"
$files = Get-SCSMObject -Class $fileAttachmentClass | Sort-Object lastmodified -Descending

#examine each file
foreach ($file in $files)
{
    #Split every File Attachment Display Name based on illegal characters
    $cleansedName = $file.DisplayName.Split([IO.Path]::GetInvalidFileNameChars())

    #if the split was successful, we'll have an array of at least 2 or more strings
    if ($cleansedName.Count -gt 1)
    {
        #Show the related Work Item
        Get-SCSMRelationshipObject -ByTarget $file | Where-Object {$_.relationshipid -eq "aa8c26dc-3a12-5f88-d9c7-753e5a8a55b4"} | Select-Object sourceobject -ExpandProperty sourceobject
        
        #Update and Show the File Attachment's new DisplayName property
        Set-SCSMObject -SMObject $file -Property "DisplayName" -Value ($cleansedName -join "") -PassThru
    }
}
```

A quick sanity check revealed this works as intended. **_But_** when I began to check the results of this in further detail (console + portal + more powershell), I discovered there is an issue present here that isn't entirely obvious. Given what you've laid out, when the attached item is an Outlook object (email/calendar appointment)...

![image](https://user-images.githubusercontent.com/6636040/139598154-65d0c80f-965a-419f-b76a-4514eeccffbf.png)

This presents an issue _only_ when  you open the file in the console as the console uses the Extension property to drive how Windows opens the file. The portal doesn't have this problem because it just uses the entire file name rather than relying on the Extension property. It's for this reason, the console crashes as you originally cite. Why I don't experience this is not clear to me but this nuance is something that was brought to my attention on #255.

This said, I'm going to need some time to re-evaluate the entire Attach-FileToWorkItem function and it's various paths with different attachment types (files, items, encrypted, signed, etc.) as the fix is really two parts:
- [x] Removing illegal characters from the File Name
- [x] Setting the correct File Extension

**Comment by infiniterecursion** [comment]
Interesting write up, I believe I've experienced this before as well but couldn't understand what was happening. I would also add to your test case attaching an Outlook object to an email where the attached Outlook object was already related to a work item in a previous email.

**Comment by AdhocAdam** [comment]
I've updated the Pull Request related to this issue seen above. It's based off of v3.4 with the sole change of handling illegal characters _and_ correctly setting the File Extension. My initial sanity checks appear to be passing in that regardless of the type of file I throw at it: 
- The DisplayName (FileName) is correctly being set
- The Extension is correctly being set
- Illegal Characters are being removed from File Names before committing into SCSM

![fileattachments](https://user-images.githubusercontent.com/6636040/141238993-528a85fa-1cd3-4b7c-ae29-e72d195aafeb.png)

If @lakam99 or @infiniterecursion is available to test the change out it would be much appreciated.

**Comment by AdhocAdam** [comment]
Just following up if @lakam99 or @infiniterecursion had a chance to test out the related pull request?

**Comment by lakam99** [comment]
Hey @AdhocAdam, sorry for the delay. I'll test this out later today!

**Comment by AdhocAdam** [comment]
Excellent!

**Comment by lakam99** [comment]
I was expecting this to work based on your tests,  but I've been unable so far to have the file attachment titled, "Hey ! ? : did you | eat the last >%%% cookie?! : | " stripped of _any_ of its illegal characters. It always comes in as if no change has been made to the code base. The connector is run through a powershell task, I've made sure its pointing to the updated version. Regardless, I will recreate the entire task tomorrow to eliminate that possibility of failure.

**Comment by AdhocAdam** [comment]
🤔 

I just tried your exact example per your original repro steps (new email, attach, choose an Outlook item from the Sent folder) and the illegal characters are stripped from the file name. Clicking on it in the portal successfully prompts for where I want to save it.

![image](https://user-images.githubusercontent.com/6636040/144144106-7c3bf0eb-bb59-4c4e-a5ac-98eb077528e8.png)

**Before**: Hey ! ? : did you | eat the last >%%% cookie?! : | 
**After**: Hey !   did you  eat the last %%% cookie!

**Comment by lakam99** [comment]
Hey @AdhocAdam, I knew your fix should solve the problem, so I figured it was an issue with our setup. Sure enough, the current SMLets build causes our dev exchange server to throw version errors. I reverted the change and manually changed the code base to reflect your changes and it successfully removed the illegal characters :) 
Thank you!

**Comment by AdhocAdam** [comment]
This fix was pulled into master via #331 and the corresponding [v3.4.1 release](https://github.com/AdhocAdam/smletsexchangeconnector/releases/tag/v3.4.1.2)
